test/e2e/targetcleanup/target_cleanup_test.go

```diff
@@ -0,0 +1,50 @@
+package targetcleanuptest
+
+import (
+	"os"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Testing cleaning up dir targets", func() {
+	When("A rebuild is done", func() {
+		It("removes and reload from the cache the dir target", func() {
+			useBobfile("with_dir_target")
+			defer releaseBobfile("with_dir_target")
+
+			b, err := BobSetup()
+			Expect(err).NotTo(HaveOccurred())
+
+			err = b.Build(ctx, "build")
+			Expect(err).NotTo(HaveOccurred())
+
+			dirContents, err := contentsOfDir("sub-dir")
+			Expect(err).NotTo(HaveOccurred())
+			Expect(dirContents).To(HaveLen(1))
+			Expect(dirContents).To(ContainElement("non-empty-file"))
+
+			// we create an empty file inside sub-dir
+			emptyFile, err := os.Create("./sub-dir/empty-file")
+			Expect(err).NotTo(HaveOccurred())
+			err = emptyFile.Close()
+			Expect(err).NotTo(HaveOccurred())
+
+			dirContents, err = contentsOfDir("sub-dir")
+			Expect(err).NotTo(HaveOccurred())
+			Expect(dirContents).To(HaveLen(2))
+			Expect(dirContents).To(ContainElement("non-empty-file"))
+			Expect(dirContents).To(ContainElement("empty-file"))
+
+			// re-build
+			err = b.Build(ctx, "build")
```

Comment:
It's not clear what's happening here. The rebuild will use the cache. Would be nice to have another test with cache disabled so we can be sure a actual rebuild is happening.